drivers: flash: sam: Redesign controller to fully utilize flash page layout

bjarki-andreasen opened the pull request:
This PR redesigns the atmel sam flash controller to fully utilize the flash page layout and capabilities of each individual soc.

An example is the ATSAM4E series socs, they have a flash layout like:
```
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 48 Kbytes          |  erase block size = 4096
|--------------------|
| 64 Kbytes          |  erase block size = 4096
|--------------------|
| ...                |

and a write block size of 8 bytes.
```
This PR adds a new `flash-pages` cell property which is used to describe this layout fully:
```
flash0: flash@400000 {
	compatible = "atmel,sam-flash", "soc-nv-flash";
	write-block-size = <8>;
	flash-pages = <&eefc 8 2048>, <&eefc 252 4096>;
};
```
Comparing this to the ATSAME70 series
```
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 112 Kbytes         |  erase block size = 8192
|--------------------|
| 128 Kbytes         |  erase block size = 8192
|--------------------|
| ...                |

and a write block size of 16 bytes.
```
It is described as
```
flash0: flash@400000 {
	compatible = "atmel,sam-flash", "soc-nv-flash";
	write-block-size = <16>;
	flash-pages = <&eefc 8 2048>, <&eefc 252 8192>;
};
```
The redesigned driver is based on the `flash_page_layout.c` driver to avoid duplicate code. The redesigned flash driver is still a bit naive, it will be expanded with logging and error handling.

Commit:
atmel: sam: flash: Add flash pages property

This commit adds layput page cells to the atmel sam flash
controller and the flash node. These allow for describing
the actual flash page layout of each soc, allowing the
flash driver to fully utilize the capabilities of the
flash.

With this update, we unlock the following capabilties:
  - utilize 2048 erase block size for small sectors
  - utilize 16384 erase block size for large sectors

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
flash: sam: Rewrite driver to dyncamically adapt to pages

This commit updates the driver to use the flash layout pages,
rewriting it to utilize the flash_page_layout.c driver to
avoid duplicate code.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Laczen commented:
Hi @bjarki-trackunit, in other flash drivers this is done by creating the correct page_layout array. For the atmel sam flash there would be two entries: one entry for the smaller pages and one for the larger pages.

bjarki-trackunit commented:
> Hi @bjarki-trackunit, in other flash drivers this is done by creating the correct page_layout array. For the atmel sam flash there would be two entries: one entry for the smaller pages and one for the larger pages.

That is what the driver is doing internally :)
https://github.com/zephyrproject-rtos/zephyr/blob/550e9ddd81fe23b5d73f7bed237df1c3803ae1f2/drivers/flash/flash_sam.c#L319-L333
Its using the new cells in the devicetree to create the proper page_layout array :)

Laczen commented:
@bjarki-trackunit, is the proposal to move the definition of flash layout from the driver to the dts ?

It would be needed to change "flash-pages" to "erase-blocks".

What is the meaning of "&eefc" ?

bjarki-trackunit commented:
> @bjarki-trackunit, is the proposal to move the definition of flash layout from the driver to the dts ?
Correct, just like `write-block-size` and `read-block-size` (which is essentially the erase block size).
> It would be needed to change "flash-pages" to "erase-blocks".
I agree `erase-blocks` :)
> What is the meaning of "&eefc" ?
Its the node label of the flash-controller :)

nandojve commented on dts/bindings/flash_controller/atmel,sam-flash-controller.yaml:
```diff
@@ -10,3 +10,11 @@ include: flash-controller.yaml
 properties:
   clocks:
     required: true
+
+  "#flash-page-cells":
+    type: int
+    const: 2
```
It will be necessary the documentation with examples. Remember that devicetree should be self explanatory and readers should not go to the datasheet to have a full understanding.

bjarki-trackunit replied:
Agreed

nandojve commented on dts/bindings/mtd/atmel,sam-flash.yaml:
```diff
@@ -0,0 +1,18 @@
+description: Atmel SAM flash area
+
+compatible: "atmel,sam-flash"
```
provide full documentation with examples

bjarki-trackunit replied:
Agreed, wanted to get some feedback first :)

bjarki-trackunit replied:
See https://github.com/zephyrproject-rtos/zephyr/compare/550e9ddd81fe23b5d73f7bed237df1c3803ae1f2..1c485944b0a69e4e06bc205ab0eee554ab3a2b03

nandojve reviewed (changes requested):
Hi @bjarki-trackunit ,

Nice!

It will took time from my side to do a proper review and test the `flash_sam.c`.

bjarki-trackunit commented:
I added the erase-blocks property and new bindings file to `sam4e`, `sam4s`, `same70`, and `samv71` series .dtsi files.

pdgendt commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,355 @@
 
-	if (len && ((offset + len - 1) > (CONFIG_FLASH_SIZE * 1024))) {
-		return false;
+	if ((offset > 0) && ((offset + len) <= config->area_size)) {
```
Why does `offset > 0`?

bjarki-trackunit replied:
Changed to `offset < 0`

pdgendt commented:
@bjarki-trackunit any update on this PR?

de-nordic commented on dts/bindings/mtd/atmel,sam-flash.yaml:
```diff
@@ -0,0 +1,109 @@
+
+  Notes:
+    The flash area layout node flash0 should have both this
+    compatible, "atmel,sam-flash", and the "soc-nv-flash"
```
mcuboot->MCUboot. And I am quite sure the MCUboot picks the flash device by `zephyr,code-partition`. The `soc-nv-flash` compatible may be used by some linker scripts, but that happens withing Zephyr.
Have you tried to build for example "Hello world", without MCUboot, and with that compatible to assigned to the node?

bjarki-trackunit replied:
I will try, my findings here are based on a samplesize of 1 :)

de-nordic replied:
OK, update on that. The "soc-nv-flash" is used by flash map macros to figure out which device actually is flash device.
If your driver makes sub-devices of controller real devices then lack of "soc-nv-flash" will work fine.
I have described the details here: https://github.com/zephyrproject-rtos/zephyr/pull/64089
In general when you try to do something to a flash device that hangs from flash-controller you have to do:
```
flash_<action>(DEVICE_DT_GET(DT_PARENT(DT_NODELABEL(flash0)))
```
rather than:
```
flash_<action>(DEVICE_DT_GET(DT_NODELABEL(flash0))
```
because flash-controller is a device and the flash0 thing hanging from it is just a configuration (?),
Flash Map needs such devices to be marked as "soc-nv-flash" to be able to get proper device pointer. We should probably have some other compatible string to address the issue.
The macro that is problematic is here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/devicetree/fixed-partitions.h#L86

de-nordic commented on dts/arm/atmel/sam4e16e.dtsi:
```diff
@@ -12,6 +12,7 @@
 		flash-controller@400e0a00 {
 			flash0: flash@400000 {
 				reg = <0x00400000 DT_SIZE_K(1024)>;
+				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
```
This is actually quite interesting. I wonder whether we could have some DT macro to directly obtain page size from given index.

bjarki-trackunit replied:
I was thinking the same, I would like to have the erase block phandles implemented as part of the `flash-controller.yaml` and have some proper macros for getting the pages if there is enough interest :)

de-nordic replied:
AT45 type devices could also use it. because they can have smaller pages at the beginning.
There are also STM32 devices that have different layouts, but this may be hard to do in DTS (I went through reading a lot of stm32 datasheets for this https://github.com/zephyrproject-rtos/zephyr/pull/36484).

Laczen replied:
I like adding the information about the erase-block sizes to the dts. I am unsure if this should be part of the `flash-controller.yaml` as also spi devices might have this property. Does the erase-blocks need a phandle to the flash-controller ? I am also unsure if it is needed to incorporate the page-layout to avoid any overlap, the page layout is the most "generic" solution supporting erase-block sizes that "grow" and "reduce" in size, while the driver knows that it are only the first pages that are smaller. The driver could be smarter than just using the page_layout.

Laczen replied:
To support also the stm32 devices the erase-blocks could be defined as `<offset blocksize blockcount>`, this could also simplify the verification of the area.

bjarki-trackunit replied:
@Laczen The phandle is required for the phandle-arrays, the phandle describes the contents of the array :) Regarding the `<offset blocksize blockcount>`, a flash area is contiguous, starting at offset 0, if there are multiple flash areas, they will get there own `erase-blocks`
```
		flash-controller@400e0a00 {
			flash0: flash@400000 {
				reg = <0x00400000 DT_SIZE_K(1024)>;
				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
			};
			flash1: flash@410000 {
				reg = <0x00410000 DT_SIZE_K(1024)>;
				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
			};
			...
```

bjarki-trackunit replied:
@de-nordic We could create a new bindings file in `bindings/mtd` named `mtd-page-layout.yaml` or similar that includes the `erase-blocks` phandle-array which can then be included by flash and at45 based devices

Laczen replied:
@bjarki-trackunit It seems that the phandle-array is not really a good fit for describing the erase-blocks as it needs a phandle :). At the moment the stm32 devices that have a "discontiguous" flash are defined as one flash node (flash0), and the page_layout is one array. The `offset` would be used to allow a discontiguity to be described.

bjarki-trackunit replied:
@Laczen The phandle to the flash controller allows you to define the contents of the `erase-blocks` to match that controller.

In atmel-sam-flash-controller.yaml
```
  "#erase-block-cells":
    type: int
    const: 2

erase-block-cells:
  - pages-count
  - pages-size
```
which is why `erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;` becomes
```
#define ... IDX_0_pages_count 8
#define ... IDX_0_pages_size 2048
#define ... IDX_1_pages_count 252
#define ... IDX_1_pages_size 4096
```
The controller (phandle) defines the specific contents of the property. If the STM32 requires the offset as well, it can define it as:
```
  "#erase-block-cells":
    type: int
    const: 3

erase-block-cells:
  - offset
  - pages-count
  - pages-size
```
Which allows you to define
```
`erase-blocks = <&flash 0 8 2048>, <&flash 0x4000 252 4096>;`
```
while still being compatible with the pages_size and pages_count macro :)

I still believe you should define multiple flash areas if there are multiple flash areas, it feels like the offset is a workaround to me :)

Commit:
drivers: flash: sam: Add logging

Add log messages to help diagnose and understand flash
driver behavior.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
drivers: flash: sam: Use interrupt to sync

Use interrupt to wait for flash controller to finish
its command and become ready.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

bjarki-trackunit commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,355 @@
 	}
 
-	flash_sam_sem_give(dev);
+	/* Checki if we've reached the end of pages to erase */
```
Applied

bjarki-trackunit commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,355 @@
 
-	flash_sam_sem_give(dev);
+	/* Checki if we've reached the end of pages to erase */
+	if (info->start_offset == erase_data->section_end) {
```
Applied

bjarki-trackunit commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,355 @@
 
-	k_sem_give(&data->sem);
+	while (!(regs->EEFC_FSR & EEFC_FSR_FRDY)) {
+		k_msleep(20);
```
Updated to use ISR to wait for operation complete

pdgendt commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
+/*
+ * This driver defines a page as the erase_block_size.
+ * This driver defines a write page as defined by the flash controller
+ * This driver defines a section as a contiguos array of bytes
```
```suggestion
 * This driver defines a section as a contiguous array of bytes
```

pdgendt commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
 };
 
-static int flash_sam_write_protection(const struct device *dev, bool enable);
+static void sam_flash_mask_ready_interrupt(const struct sam_flash_config *config)
```
```suggestion
static inline void sam_flash_mask_ready_interrupt(const struct sam_flash_config *config)
```

pdgendt commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
+}
 
-static inline void flash_sam_sem_take(const struct device *dev)
+static void sam_flash_unmask_ready_interrupt(const struct sam_flash_config *config)
```
```suggestion
static inline void sam_flash_unmask_ready_interrupt(const struct sam_flash_config *config)
```

pdgendt reviewed (dismissed):
Some minor nits, looks good

bjarki-trackunit commented:
@nandojve Could you find time to revisit this one?

nandojve commented:
> @nandojve Could you find time to revisit this one?

Hi @bjarki-trackunit ,

Maybe on the weekend. I need to take time to test on HW.

pdgendt commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
-	if (len && ((offset + len - 1) > (CONFIG_FLASH_SIZE * 1024))) {
-		return false;
+	if (regs->EEFC_FSR & EEFC_FSR_FCMDE) {
+		LOG_ERR("Invalid command requested");
+		return -EPERM;
 	}
 
-	return true;
+	if (regs->EEFC_FSR & EEFC_FSR_FLOCKE) {
+		LOG_ERR("Tried to modify locked region");
+		return -EPERM;
+	}
+
+	if (regs->EEFC_FSR & EEFC_FSR_FLERR) {
+		LOG_ERR("Programming failed");
+		return -EPERM;
+	}
```
Flags in `EEFC_FSR` are cleared when read.

```c
	uint32_t status = regs->EEFC_FSR;

	if (status & EEFC_FSR_FCMDE) {
		LOG_ERR("Invalid command requested");
		return -EPERM;
	}
	// ...
```

pdgendt replied:
@bjarki-trackunit this still needs to be fixed IMO

bjarki-trackunit replied:
Fixed with
https://github.com/zephyrproject-rtos/zephyr/blob/a06dfce570b5f7a56c75cf757212e9455bd375ad/drivers/flash/flash_sam.c#L100-L136

nandojve commented:
Hi @bjarki-trackunit ,

I'll test today. Can you apply missing findings from @pdgendt ?

nandojve commented:
BTW: This driver should pass on CI.

How are you testing this because I can not build tests/drivers/common?
I'll not accept any changes in drivers that restructure everything without proper test.

nandojve commented:
Hi @bjarki-trackunit ,

I think it will be necessary introduce a test that allow us to run from a high address memory and validate layouts that you mentioned.

```
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 48 Kbytes          |  erase block size = 4096
|--------------------|
| 64 Kbytes          |  erase block size = 4096
|--------------------|
| ...                |

|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 8 Kbytes           |  erase block size = 2048
|--------------------|
| 112 Kbytes         |  erase block size = 8192
|--------------------|
| 128 Kbytes         |  erase block size = 8192
|--------------------|
| ...                |
```

It will be necessary to stress the driver to make sure it works fine when crossing block boundaries too.

pdgendt commented:
Ping @bjarki-trackunit :)

bjarki-andreasen committed:
tests: flash: Add erase_blocks test for flash controllers

Add test suite which tests writing to and erasing all pages,
of any size defined by the page layout.

If the test is built with MCUBOOT, the test will jump into the
application, then run write/erase/boundary tests on the
bootload partition. If MCUBOOT is not selected, the test will
target the slot1 partition instead, performing the same tests.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
### Update
* Added test suite to test writing to and erasing all flash pages in the boot partition. Validated using b_u585i_iot02a board
* Working on getting it to work with the ATSAM4E16C which explodes if used with mcuboot for some reason...

pdgendt commented:
> ### Update
> * Added test suite which writes to and erases all flash pages in the boot partition. Validated using b_u585i_iot02a board
> * Working on getting it to work with the ATSAM4E16C which explodes if used with mcuboot for some reason...

The `sam4s` also needed changes to get `mcuboot` to work, see https://github.com/zephyrproject-rtos/zephyr/pull/66499

bjarki-trackunit commented:
@pdgendt I will definitely give it a go!

bjarki-trackunit commented:
With the fix from this PR #66499 applied to the sam4e, the new flash page layout test works like a charm with this PR's driver and an ATSAM4E16C SOC. This is the produced output:
```
Running TESTSUITE flash_page_layout
===================================================================
START - test_erase_single_pages_in_boot_part
I: Erasing bootloader section of size 65536 at offset 0 controlled by flash-controller@400e0a00
I: Writing across page boundary at 2048, between page index 0 and 1
I: Writing across page boundary at 4096, between page index 1 and 2
I: Writing across page boundary at 6144, between page index 2 and 3
I: Writing across page boundary at 8192, between page index 3 and 4
I: Writing across page boundary at 10240, between page index 4 and 5
I: Writing across page boundary at 12288, between page index 5 and 6
I: Writing across page boundary at 14336, between page index 6 and 7
I: Writing across page boundary at 16384, between page index 7 and 8
I: Writing across page boundary at 20480, between page index 8 and 9
I: Writing across page boundary at 24576, between page index 9 and 10
I: Writing across page boundary at 28672, between page index 10 and 11
I: Writing across page boundary at 32768, between page index 11 and 12
I: Writing across page boundary at 36864, between page index 12 and 13
I: Writing across page boundary at 40960, between page index 13 and 14
I: Writing across page boundary at 45056, between page index 14 and 15
I: Writing across page boundary at 49152, between page index 15 and 16
I: Writing across page boundary at 53248, between page index 16 and 17
I: Writing across page boundary at 57344, between page index 17 and 18
I: Writing across page boundary at 61440, between page index 18 and 19
I: Erasing page at 0 of size 2048 with index 0
I: Erasing page at 2048 of size 2048 with index 1
I: Erasing page at 4096 of size 2048 with index 2
I: Erasing page at 6144 of size 2048 with index 3
I: Erasing page at 8192 of size 2048 with index 4
I: Erasing page at 10240 of size 2048 with index 5
I: Erasing page at 12288 of size 2048 with index 6
I: Erasing page at 14336 of size 2048 with index 7
I: Erasing page at 16384 of size 4096 with index 8
I: Erasing page at 20480 of size 4096 with index 9
I: Erasing page at 24576 of size 4096 with index 10
I: Erasing page at 28672 of size 4096 with index 11
I: Erasing page at 32768 of size 4096 with index 12
I: Erasing page at 36864 of size 4096 with index 13                                                                         
I: Erasing page at 40960 of size 4096 with index 14                                                                         
I: Erasing page at 45056 of size 4096 with index 15                                                                         
I: Erasing page at 49152 of size 4096 with index 16                                                                         
I: Erasing page at 53248 of size 4096 with index 17                                                                         
I: Erasing page at 57344 of size 4096 with index 18                                                                         
I: Erasing page at 61440 of size 4096 with index 19                                                                         
 PASS - test_erase_single_pages_in_boot_part in 1.818 seconds                                                               
===================================================================                                                         
START - test_write_across_page_boundaries_in_boot_part                                                                      
I: Erasing bootloader section of size 65536 at offset 0 controlled by flash-controller@400e0a00                             
I: Writing across page boundary at 2048, between page index 0 and 1
I: Writing across page boundary at 4096, between page index 1 and 2
I: Writing across page boundary at 6144, between page index 2 and 3
I: Writing across page boundary at 8192, between page index 3 and 4
I: Writing across page boundary at 10240, between page index 4 and 5
I: Writing across page boundary at 12288, between page index 5 and 6
I: Writing across page boundary at 14336, between page index 6 and 7
I: Writing across page boundary at 16384, between page index 7 and 8
I: Writing across page boundary at 20480, between page index 8 and 9
I: Writing across page boundary at 24576, between page index 9 and 10
I: Writing across page boundary at 28672, between page index 10 and 11
I: Writing across page boundary at 32768, between page index 11 and 12
I: Writing across page boundary at 36864, between page index 12 and 13
I: Writing across page boundary at 40960, between page index 13 and 14
I: Writing across page boundary at 45056, between page index 14 and 15
I: Writing across page boundary at 49152, between page index 15 and 16
I: Writing across page boundary at 53248, between page index 16 and 17
I: Writing across page boundary at 57344, between page index 17 and 18
I: Writing across page boundary at 61440, between page index 18 and 19
 PASS - test_write_across_page_boundaries_in_boot_part in 0.967 seconds
===================================================================
TESTSUITE flash_page_layout succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [flash_page_layout]: pass = 2, fail = 0, skip = 0, total = 2 duration = 2.785 seconds
 - PASS - [flash_page_layout.test_erase_single_pages_in_boot_part] duration = 1.818 seconds
 - PASS - [flash_page_layout.test_write_across_page_boundaries_in_boot_part] duration = 0.967 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

bjarki-trackunit commented:
@nandojve Regarding failing to build the flash common test suite, you have to add the following
```
&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		storage_partition: partition@f0000 {
			label = "storage";
			reg = <0xf0000 0xa000>;
		};
	};
};
```
to make it build for your board.
For my custom board with a ATSAM4E16C, I get the following output
```
Running TESTSUITE flash_driver
===================================================================
START - test_read_unaligned_address
 PASS - test_read_unaligned_address in 0.007 seconds
===================================================================
TESTSUITE flash_driver succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [flash_driver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.007 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.007 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

nandojve commented:
Hi @bjarki-trackunit ,

First of all, I really appreciate your changes and the effort you are expending to make this. I think your contribution is very valuable and I want to have it ASAP.

Please, add the suggested layout to common tests for the _sam_v71_xult_. It is important we make it available. There is a bug in the test BTW.

```
*** Booting Zephyr OS build zephyr-v3.5.0-3223-g37ca48d8f215 ***
Running TESTSUITE flash_driver
===================================================================
START - test_read_unaligned_address

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/flash/common/src/main.c:128: flash_driver_test_read_unaligned_address: (memcmp(buf)
Flash read failed at len=1, ad_o=0, buf_o=1
 FAIL - test_read_unaligned_address in 0.022 seconds
===================================================================
TESTSUITE flash_driver failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -   0.00% [flash_driver]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.022 seconds
 - FAIL - [flash_driver.test_read_unaligned_address] duration = 0.022 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```

Remember that SAM requires 8 byte alignment.

```diff
diff --git a/tests/drivers/flash/common/src/main.c b/tests/drivers/flash/common/src/main.c
index 79d75ca9b1..15f3c76b75 100644
--- a/tests/drivers/flash/common/src/main.c
+++ b/tests/drivers/flash/common/src/main.c
@@ -47,7 +47,7 @@
 
 static const struct device *const flash_dev = TEST_AREA_DEVICE;
 static struct flash_pages_info page_info;
-static uint8_t __aligned(4) expected[EXPECTED_SIZE];
+static uint8_t __aligned(8) expected[EXPECTED_SIZE];
 
 static void *flash_driver_setup(void)
 {
```
So, after the patch
```
*** Booting Zephyr OS build zephyr-v3.5.0-3223-g37ca48d8f215 ***
Running TESTSUITE flash_driver
===================================================================
START - test_read_unaligned_address
 PASS - test_read_unaligned_address in 0.002 seconds
===================================================================
TESTSUITE flash_driver succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [flash_driver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.002 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.002 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

```

I see that now the newer test depends on MCUboot which complicate the test for all boards.

1- Is it possible to add two distincts configurations to enable the build with MCUboot and with normal Zephyr application? I think we should think about this because usually people only test one mode and not both. So you may need [2]
2- Should we add support to build and run from SRAM to allow a more elaborated test? CC: @pdgendt did you already tried ?
3- Maybe you can send a suggestion to flash layout for SAM4E and we enabled MCUboot support to it.

Why:

I just tried to build and run for samv71 and it stuck here (I'm not using MCUboot on purpose):
```
*** Booting Zephyr OS build zephyr-v3.5.0-3223-g37ca48d8f215 ***
Running TESTSUITE flash_page_layout
===================================================================
START - test_erase_single_pages_in_boot_part
I: Erasing bootloader section of size 65536 at offset 0 controlled by flash-controller@400e0c00
```

If I try to build SAM4E I get a lot of building errors because there is no flash layout available.

bjarki-trackunit commented:
@nandojve
1. Yes, I can make it use the slot2 partition instead of the boot partition if MCUBoot is not used, this does not allow us to test the first smaller pages in most cases since z_test is quite large in ROM. I think both is actually the best solution, first test without MCUBoot to test if the flash driver is working at all, then with MCUBoot to really stress test it (MCUBoot will not boot if the flash driver doesn't work)
2. This would complicate the test immensely IMO, running from RAM is not exactly common compared to running with MCUBoot from flash. So I would say no, the two tests with and without MCUBoot should do :)
3. I agree we should do this in the future, but it is way outside of the scope of this PR

Regarding the test without MCUBoot:
You can't run the new test without MCUBoot :) The tests stops here
```
I: Erasing bootloader section of size 65536 at offset 0 controlled by flash-controller@400e0c00
```
because the start of the application is being erased by the test suite (MCUBoot is supposed to be there).

Using MCUBoot is pretty much the simplest way to jump to the application from the start of the flash (the smaller pages), and it is a generic solution. The tests immediately work for all nrf and stm boards I have laying around, it will work for the sam based boards quite soon as well (it does for mine with the early fix from #66499)

nandojve commented:
> @nandojve
> 
> 1. Yes, I can make it use the slot2 partition instead of the boot partition if MCUBoot is not used, this does not allow us to test the first smaller pages in most cases since z_test is quite large in ROM. I think both is actually the best solution, first test without MCUBoot to test if the flash driver is working at all, then with MCUBoot to really stress test it (MCUBoot will not boot if the flash driver doesn't work)
> 2. This would complicate the test immensely IMO, running from RAM is not exactly common compared to running with MCUBoot from flash. So I would say no, the two tests with and without MCUBoot should do :)
> 3. I agree we should do this in the future, but it is way outside of the scope of this PR
> 
> Regarding the test without MCUBoot: You can't run the new test without MCUBoot :) The tests stops here
> 
> ```
> I: Erasing bootloader section of size 65536 at offset 0 controlled by flash-controller@400e0c00
> ```
> 
> because the start of the application is being erased by the test suite (MCUBoot is supposed to be there).
> 
> Using MCUBoot is pretty much the simplest way to jump to the application from the start of the flash (the smaller pages), and it is a generic solution. The tests immediately work for all nrf and stm boards I have laying around, it will work for the sam based boards quite soon as well (it does for mine with the early fix from #66499)
> 
> For whichever board you are using which is based on the SAM4E, you need to include the partition layout, just like it is for the `sam_v71_xult` and most other boards :)

Hi @bjarki-trackunit ,

Can you update PR using your [1] suggestion?

nandojve commented:
ping

bjarki-trackunit commented:
### Update

The page_layout test suite now either targets the bootloader partition or the slo1 partition depending on whether MCUBOOT is selected.

Laczen commented:
@bjarki-trackunit, some devices have a "gap" inside the pages definition that cannot be used. Would it be possible to include a offset in the `erase-blocks` definition, this allows defining such a gap.

bjarki-trackunit commented:
> @bjarki-trackunit, some devices have a "gap" inside the pages definition that cannot be used. Would it be possible to include a offset in the `erase-blocks` definition, this allows defining such a gap.

While it is possible, I would suggest creating multiple flash areas instead, as a flash area is contiguous. Like
```
flash-controller@deaddead {
        flash0: flash@0 {
                /* ends at 0x80000 */
                reg = <0x0 DT_SIZE_K(512)>;
        };

        /* gap is here */

        flash1: flash@0x90000 {
                reg = <0x90000 DT_SIZE_K(512)>;
        };
};
```

Laczen commented:
> > @bjarki-trackunit, some devices have a "gap" inside the pages definition that cannot be used. Would it be possible to include a offset in the `erase-blocks` definition, this allows defining such a gap.
> 
> While it is possible, I would suggest creating multiple flash areas instead, as a flash area is contiguous. Like
> 
> ```
> flash-controller@deaddead {
>         flash0: flash@0 {
>                 /* ends at 0x80000 */
>                 reg = <0x0 DT_SIZE_K(512)>;
>         };
> 
>         /* gap is here */
> 
>         flash1: flash@0x90000 {
>                 reg = <0x90000 DT_SIZE_K(512)>;
>         };
> };
> ```
> 
> This may not be possible looking at the flash API though, it seems there is only one page layout per controller, not per area. We may need to update the flash API to support either multiple areas or gaps (though I do lean towards multiple areas)
> 
> That said, I don't know if it is feasible to "hide" the gap in the flash area without a bunch of linker magic, if one was to write an application that spans such a gap, this would break the application, right? since it is reading flash directly, not through the flash API which can hide such gaps

You can disregard my request, this gap is not visible outside of the driver, the driver does the mapping from a continuous address region defined in the dts to a discontinuous address region in the driver. The mapping could be specified in the dts using ranges.

nandojve reviewed (dismissed):
Hi @bjarki-trackunit ,

I tested on sam_v71_xult mainline is working. After I added the #66499 sam4e_xpro and sam4s_xplained worked too. However, I found a possible regression on #66499 for samv71.

Besides that, this is ready to go!

Thank you so much for this job!

Laczen commented:
@bjarki-trackunit, regarding the `&eefc`, did you need it for dts processing or does it have a real meaning. What should be used for external flash?

Laczen reviewed (dismissed):
Nice job, thanks.

Laczen commented:
> For external flash, it should be a flash controller, with a flash area, just like internal flash :)
> 
Could it not just be part of the device yaml?

bjarki-trackunit commented:
> > For external flash, it should be a flash controller, with a flash area, just like internal flash :)
> 
> Could it not just be part of the device yaml?

Well, the device is a flash controller, right?

Laczen commented:
@bjarki-trackunit, maybe a future addition could be to fall back to an `erase-block-size` based flash page layout (the pre PR layout) when `erase-blocks` is not specified.

de-nordic commented:
> > > For external flash, it should be a flash controller, with a flash area, just like internal flash :)
> > 
> > 
> > Could it not just be part of the device yaml?
> 
> Well, the device is a flash controller, right?

Unfortunately.

de-nordic commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
+	sam_data->erase_data.succeeded = false;
+	flash_page_foreach(dev, sam_flash_erase_foreach_page, sam_data);
+	if (!sam_data->erase_data.succeeded) {
+		k_spin_unlock(&sam_data->lock, key);
```
you could probably have the unlock above if and discard the one below.

bjarki-trackunit replied:
The unlock needs to be after since the `erase_data` context is shared, and must be protected until read.

de-nordic reviewed (dismissed):
Some nits, but generally OK.

bjarki-trackunit commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
+	const struct sam_flash_config *config = dev->config;
 
-	Efc * const efc = config->regs;
+	if ((offset >= 0) && ((offset + len) <= config->area_size)) {
+		return true;
```
Added a check to validate the addition of offset and len are safe
https://github.com/zephyrproject-rtos/zephyr/blob/84b6ed90d289597e90954bb0637ed8db3d7a0ed3/drivers/flash/flash_sam.c#L55-L66

bjarki-trackunit commented on drivers/flash/flash_sam.c:
```diff
@@ -1,404 +1,492 @@
-		 * might be set outside of the programming code.
-		 */
+	if (((offset % config->parameters.write_block_size) == 0) &&
+	    ((len % config->parameters.write_block_size) == 0)) {
```
Changed to use binary operations as the value will always be aligned with an exponent of base 2
https://github.com/zephyrproject-rtos/zephyr/blob/84b6ed90d289597e90954bb0637ed8db3d7a0ed3/drivers/flash/flash_sam.c#L70C1-L71C2

pdgendt reviewed (dismissed):
Thanks! Tested on sam4s.

Laczen commented:
@bjarki-trackunit, I have some annoying remark.

The PR adds a test `page_layout` that erases and writes to flash. A `page_layout` test should focus on the routines that are provided by the `page_layout` api: `flash_get_page_info_by_offs()`, `flash_get_page_info_by_idx()` and `flash_page_foreach`. The correct functionality for each of these routines should be tested and if possible without any erase/write. Also this test should be carried out in userspace if possible (at the moment the `flash_page_foreach` is not supported for userspace access).

Could the test be renamed into something else so that can leave the name `page_layout` for the test of the `page_layout` api?

bjarki-trackunit commented:
> Could the test be renamed into something else so that can leave the name `page_layout` for the test of the `page_layout` api?

sure, I have no strong feeling on the name, I could name it erase_write_boundary or just erase_write_validate, or maybe you have a good name?

Laczen commented:
> > Could the test be renamed into something else so that can leave the name `page_layout` for the test of the `page_layout` api?
> 
> sure, I have no strong feeling on the name, I could name it erase_write_boundary or just erase_write_validate, or maybe you have a good name?

What about `erase_blocks` ?

bjarki-trackunit commented:
renamed `page_layout` test to `erase_blocks`

Laczen commented:
> renamed `page_layout` test to `erase_blocks`

Maybe in the long run we should move all present flash tests under `functional` and add a separate `api` folder for api tests (that also verify the userspace access).
